File: 1/src/main.rs
use std::fs::File;
use std::io::{BufRead, BufReader, Error, ErrorKind, Result};
use std::path::Path;

fn read_file(filename: &str) -> Result<Vec<i64>> {
    let path = Path::new(&filename);

    let file = File::open(&path)?;
    let br = BufReader::new(file);
    br.lines()
        .map(|line| line.and_then(|v| v.parse().map_err(|e| Error::new(ErrorKind::InvalidData, e))))
        .collect()
}

fn main() {
    let data: Vec<i64> = read_file("src/input.txt")
        .unwrap()
        .drain(..)
        .filter(|v| *v <= 1999)
        .collect();

    // Brute force sollution
    'outer: for i in 0..data.len() {
        for j in 0..data.len() {
            let i_val = data[i];
            let j_val = data[j];
            let partial_sum = i_val + j_val;
            if i == j || partial_sum > 2020 {
                continue;
            }
            for k in 0..data.len() {
                if k == j || k == i {
                    continue;
                }
                let k_val = data[k];
                let k_partial_sum = partial_sum +  k_val;
                if k_partial_sum == 2020 {
                    println!("Found one!: {}", i_val * j_val * k_val);
                    break 'outer;
                }
            }
        }
    }
}

File: 1/cpp/src/main.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <string>

using namespace std;

/* Given a list of numbers, find 3 numbers that sum to 2020. Then multiply them.
*/

vector<int> ReadFile(const string &filename)
{
    vector<int> contents = {};
    fstream input_file;
    input_file.open(filename, ios::in);
    if (!input_file)
    {
        cout << "Error opening file!";
        return contents;
    }
    string buffer;
    while (!input_file.eof())
    {
        getline(input_file, buffer, '\n');
        try
        {
            int current_number = stoi(buffer);
            contents.push_back(current_number);
        }
        catch (exception)
        {
        }
    }
    return contents;
}

int main()
{
    vector<int> data = ReadFile("src/input.txt");

    auto data_length = data.size();
    for (int i = 0; i < data_length; ++i)
    {
        for (int j = 0; j < data_length; ++j)
        {
            int i_val = data[i];
            int j_val = data[j];
            int sum_of_two = i_val + j_val;
            if (i == j || sum_of_two > 2020)
            {
                continue;
            }
            for (int k = 0; k < data_length; ++k)
            {
                if (k == j || k == i)
                {
                    continue;
                }
                int k_val = data[k];
                int sum_of_three = sum_of_two + k_val;
                if (sum_of_three == 2020)
                {
                    long result = i_val * j_val * k_val;
                    cout << "Found one match!: " << i_val << ", " << j_val << ", " << k_val << endl;
                    cout << "Found one match!: " << result << endl;
                    return 0;
                }
            }
        }
    }
}
